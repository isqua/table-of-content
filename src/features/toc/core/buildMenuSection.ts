import type { MenuItem, PageDescriptor, PageHighlight, PageId, PageURL, SectionHighlight, TableOfContent } from '../types'

type BuildMenuBaseOptions = {
    url: PageURL
    breadcrumbs: PageDescriptor[]
    filter?: Set<PageDescriptor> | null
}

type BuildMenuTopLevelOptions = BuildMenuBaseOptions

type BuildMenuNestingOptions = BuildMenuBaseOptions & {
    parentId: string
    level: number
    highlight: SectionHighlight
}

type BuildMenuOptions = BuildMenuTopLevelOptions | BuildMenuNestingOptions

type PageProps = {
    highlight: PageHighlight
    defaultOpenState: boolean
    level: number
}

const mapPageToMenuItem = (page: PageDescriptor, props: PageProps): MenuItem => ({
    id: page.id,
    parentId: page.parentId,
    title: page.title,
    url: page.url ?? '',
    highlight: props.highlight,
    level: props.level,
    hasChildren: Boolean(page.pages?.length),
    defaultOpenState: props.defaultOpenState
})

const getPagesForParent = (toc: TableOfContent, parentId?: PageId): PageId[] => {
    if (!parentId) {
        return toc.topLevelIds
    }

    if (!(parentId in toc.entities.pages)) {
        return []
    }

    return toc.entities.pages[parentId]?.pages ?? []
}

const getPageFilter = (filter?: Set<PageDescriptor> | null) => {
    if (!filter) {
        return function(page: PageDescriptor | undefined): page is PageDescriptor {
            return page !== undefined
        }
    }

    return function(page: PageDescriptor | undefined): page is PageDescriptor {
        return page !== undefined && filter.has(page)
    }
}

export const buildMenuSection = (toc: TableOfContent, options: BuildMenuOptions): MenuItem[] => {
    const {
        url,
        parentId,
        breadcrumbs = [],
        level = 0,
        highlight: sectionHighlight,
        filter,
    } = options as BuildMenuNestingOptions

    const menu: MenuItem[] = []
    const pagesIds = getPagesForParent(toc, parentId)
    const shouldAddPage = getPageFilter(filter)
    const hasFilter = Boolean(filter)

    for (const pageId of pagesIds) {
        const page = toc.entities.pages[pageId]

        if (shouldAddPage(page)) {
            let highlight: PageHighlight = sectionHighlight
            let defaultOpenState = false

            if (page.url === url) {
                highlight = 'active'
                defaultOpenState = Boolean(page.pages?.length)
            } else if (breadcrumbs[level] === page) {
                highlight = 'parent'
                defaultOpenState = true
            }

            if (hasFilter && page.pages?.length) {
                defaultOpenState = true
            }

            menu.push(mapPageToMenuItem(page, {
                highlight,
                level,
                defaultOpenState,
            }))
        }
    }

    return menu
}
