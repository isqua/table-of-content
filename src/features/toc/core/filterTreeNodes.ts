import type { PageDescriptor, TableOfContent } from '../types'

const isTextMatch = (value: string, search: string) => {
    return value.toLowerCase().includes(search)
}

export const filterTreeNodes = (toc: TableOfContent, search: string): Set<PageDescriptor> => {
    const filterResult = new Set<PageDescriptor>()
    const normalizedSearch = search.toLocaleLowerCase()

    for (const page of Object.values(toc.entities.pages)) {
        if (page && isTextMatch(page.title.toLocaleLowerCase(), normalizedSearch)) {

            let currentPage: PageDescriptor | undefined = page

            while (currentPage && toc.entities.pages[currentPage.id] && !filterResult.has(currentPage)) {
                filterResult.add(currentPage)

                currentPage = toc.entities.pages[currentPage.parentId]
            }
        }
    }

    return filterResult
}
