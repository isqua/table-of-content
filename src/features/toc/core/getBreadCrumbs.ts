import type { PageDescriptor, PageURL, PageId, TableOfContent } from '../types'
import { getCurrentPage } from './getCurrentPage'

/**
 * @param toc The TOC data
 * @param url Page URL
 * @returns Current page and the array of its ancestors from top to bottom
 */
export const getBreadCrumbs = (toc: TableOfContent, url?: PageURL): PageDescriptor[] => {
    const breadcrumbs: PageDescriptor[] = []

    // We trust the API, but not completely
    const visitedIds = new Set<PageId>()

    if (!url) {
        return breadcrumbs
    }

    let currentPage: PageDescriptor | undefined = getCurrentPage(toc, url)

    while (currentPage && toc.entities.pages[currentPage.id] && !visitedIds.has(currentPage.id)) {
        breadcrumbs.push(currentPage)
        visitedIds.add(currentPage.id)

        currentPage = toc.entities.pages[currentPage.parentId]
    }

    return breadcrumbs.reverse()
}
