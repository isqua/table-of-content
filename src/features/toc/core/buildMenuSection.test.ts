import { describe, expect, it } from 'vitest'

import type { TableOfContent } from '../types'
import { buildMenuSection } from './buildMenuSection'
import { filterTreeNodes } from './filterTreeNodes'
import { getBreadCrumbs } from './getBreadCrumbs'

import tocFlat from '../../../test/fixtures/toc/flat.json'
import tocTwoLevels from '../../../test/fixtures/toc/two-levels.json'
import tocThreeLevels from '../../../test/fixtures/toc/three-levels.json'

describe('toc/buildMenuSection', () => {
    it('should build a menu and highlight current page', () => {
        const toc: TableOfContent = tocFlat
        const currentUrl = '/bar.html'
        const breadcrumbs = getBreadCrumbs(toc, currentUrl)
        const menu = buildMenuSection(toc, { url: currentUrl, breadcrumbs })

        expect(menu).toMatchSnapshot()
    })

    it('should build a menu even if there are no current page', () => {
        const toc: TableOfContent = tocFlat
        const currentUrl = '/'
        const breadcrumbs = getBreadCrumbs(toc, currentUrl)
        const menu = buildMenuSection(toc, { url: currentUrl, breadcrumbs })

        expect(menu).toMatchSnapshot()
    })

    it('should build an empty menu if page with parent id does not exists', () => {
        const toc: TableOfContent = tocFlat
        const currentUrl = '/'
        const parentId = 'this-id-does-not-exist'
        const breadcrumbs = getBreadCrumbs(toc, currentUrl)
        const menu = buildMenuSection(toc, { url: currentUrl, breadcrumbs, parentId })

        expect(menu).toEqual([])
    })

    it('should build an empty menu if there are no top level ids ', () => {
        const toc: TableOfContent = {
            entities: { pages: {} },
            topLevelIds: [],
        }
        const currentUrl = '/'
        const breadcrumbs = getBreadCrumbs(toc, currentUrl)
        const menu = buildMenuSection(toc, { url: currentUrl, breadcrumbs })

        expect(menu).toMatchSnapshot()
    })

    it('should show page children', () => {
        const toc: TableOfContent = tocTwoLevels
        const currentUrl = '/bar.html'
        const breadcrumbs = getBreadCrumbs(toc, currentUrl)
        const menu = buildMenuSection(toc, { url: currentUrl, breadcrumbs })

        expect(menu).toMatchSnapshot()
    })

    it('should build a nested menu', () => {
        const toc: TableOfContent = tocTwoLevels
        const currentUrl = '/bar-features.html'
        const breadcrumbs = getBreadCrumbs(toc, currentUrl)
        const menu = buildMenuSection(toc, { url: currentUrl, parentId: 'bar', breadcrumbs })

        expect(menu).toMatchSnapshot()
    })

    it('should filter pages', () => {
        const toc: TableOfContent = tocFlat
        const currentUrl = '/bar.html'
        const breadcrumbs = getBreadCrumbs(toc, currentUrl)
        const filter = filterTreeNodes(toc, 'ba')
        const menu = buildMenuSection(toc, { url: currentUrl, breadcrumbs, filter })

        expect(menu).toMatchSnapshot()
    })

    it('should expand all sections while filtering', () => {
        const toc: TableOfContent = tocThreeLevels
        const currentUrl = '/'
        const breadcrumbs = getBreadCrumbs(toc, currentUrl)
        const filter = filterTreeNodes(toc, 'features')

        const menu = [
            ...buildMenuSection(toc, { url: currentUrl, breadcrumbs, filter }),
            ...buildMenuSection(toc, { parentId: 'bar', url: currentUrl, breadcrumbs, filter }),
            ...buildMenuSection(toc, { parentId: 'bar_features', url: currentUrl, breadcrumbs, filter }),
        ]

        expect(menu).toMatchSnapshot()
    })
})
