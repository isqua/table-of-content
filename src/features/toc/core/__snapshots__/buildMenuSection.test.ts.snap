// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`toc/buildMenuSection > should build a menu and highlight current page 1`] = `
[
  {
    "defaultOpenState": false,
    "hasChildren": false,
    "highlight": "active",
    "id": "bar",
    "level": 0,
    "parentId": "root",
    "title": "Bar",
    "url": "/bar.html",
  },
  {
    "defaultOpenState": false,
    "hasChildren": false,
    "highlight": undefined,
    "id": "baz",
    "level": 0,
    "parentId": "root",
    "title": "Baz",
    "url": "/baz.html",
  },
  {
    "defaultOpenState": false,
    "hasChildren": false,
    "highlight": undefined,
    "id": "foo",
    "level": 0,
    "parentId": "root",
    "title": "Foo",
    "url": "/foo.html",
  },
]
`;

exports[`toc/buildMenuSection > should build a menu even if there are no current page 1`] = `
[
  {
    "defaultOpenState": false,
    "hasChildren": false,
    "highlight": undefined,
    "id": "bar",
    "level": 0,
    "parentId": "root",
    "title": "Bar",
    "url": "/bar.html",
  },
  {
    "defaultOpenState": false,
    "hasChildren": false,
    "highlight": undefined,
    "id": "baz",
    "level": 0,
    "parentId": "root",
    "title": "Baz",
    "url": "/baz.html",
  },
  {
    "defaultOpenState": false,
    "hasChildren": false,
    "highlight": undefined,
    "id": "foo",
    "level": 0,
    "parentId": "root",
    "title": "Foo",
    "url": "/foo.html",
  },
]
`;

exports[`toc/buildMenuSection > should build a nested menu 1`] = `
[
  {
    "defaultOpenState": false,
    "hasChildren": false,
    "highlight": undefined,
    "id": "bar_install",
    "level": 0,
    "parentId": "bar",
    "title": "Bar: Install",
    "url": "/bar-install.html",
  },
  {
    "defaultOpenState": false,
    "hasChildren": false,
    "highlight": "active",
    "id": "bar_features",
    "level": 0,
    "parentId": "bar",
    "title": "Bar: Features",
    "url": "/bar-features.html",
  },
]
`;

exports[`toc/buildMenuSection > should build an empty menu if there are no top level ids  1`] = `[]`;

exports[`toc/buildMenuSection > should expand all sections while filtering 1`] = `
[
  {
    "defaultOpenState": true,
    "hasChildren": true,
    "highlight": undefined,
    "id": "bar",
    "level": 0,
    "parentId": "root",
    "title": "Bar",
    "url": "/bar.html",
  },
  {
    "defaultOpenState": true,
    "hasChildren": true,
    "highlight": undefined,
    "id": "bar_features",
    "level": 0,
    "parentId": "bar",
    "title": "Bar: Features",
    "url": "/bar-features.html",
  },
  {
    "defaultOpenState": false,
    "hasChildren": false,
    "highlight": undefined,
    "id": "bar_features_beer",
    "level": 0,
    "parentId": "bar_features",
    "title": "Bar Features: Beer",
    "url": "/bar-features-beer.html",
  },
]
`;

exports[`toc/buildMenuSection > should filter pages 1`] = `
[
  {
    "defaultOpenState": false,
    "hasChildren": false,
    "highlight": "active",
    "id": "bar",
    "level": 0,
    "parentId": "root",
    "title": "Bar",
    "url": "/bar.html",
  },
  {
    "defaultOpenState": false,
    "hasChildren": false,
    "highlight": undefined,
    "id": "baz",
    "level": 0,
    "parentId": "root",
    "title": "Baz",
    "url": "/baz.html",
  },
]
`;

exports[`toc/buildMenuSection > should show page children 1`] = `
[
  {
    "defaultOpenState": true,
    "hasChildren": true,
    "highlight": "active",
    "id": "bar",
    "level": 0,
    "parentId": "root",
    "title": "Bar",
    "url": "/bar.html",
  },
  {
    "defaultOpenState": false,
    "hasChildren": false,
    "highlight": undefined,
    "id": "baz",
    "level": 0,
    "parentId": "root",
    "title": "Baz",
    "url": "/baz.html",
  },
  {
    "defaultOpenState": false,
    "hasChildren": false,
    "highlight": undefined,
    "id": "foo",
    "level": 0,
    "parentId": "root",
    "title": "Foo",
    "url": "/foo.html",
  },
]
`;
